### ISSUE TITLE
Expose `interp_*d` and `spread_*d` functions for more general usage?

### ISSUE THREAD

**Issue by michael-0brien** [geoffroyO/nufftax#10]
Hello again! Again, awesome package. I was wondering the following: would you be open to exposing API that would support more general usage of your `interp_*d` and `spread_*d` functions? I'd like to use these for spreading of short-ranged gaussians and other kernels, without using the full nufft pipeline.

An oversimplification of what I'm thinking is changing the following (here e.g. only for `spread_1d`):

```python
def spread_1d(
    x: jax.Array,
    c: jax.Array,
    nf: int,
    kernel_params: KernelParams,  # Specifically for nufft kernel
) -> jax.Array:
    ...
```

To something like this:

```python
def spread_1d(
    x: jax.Array,
    c: jax.Array,
    nf: int,
    kernel_fn: Callable[[Array], [Array]],  # Supports generic `f(z) = f_val`
    kernel_with_derivative_fn: Callable[[Array], [Array, Array]],  # And `f(z) = f_val, df_dz_val`
    nspread: int,
) -> jax.Array:
    ...
```

I know a request like this could be beyond the scope of the package, but I'm curious to hear what you think. I'd be happy to put together a PR.

**Comment by geoffroyO** [comment]
Hi @michael-0brien, thanks again! Exposing spread/interp for custom kernels makes sense.

The catch is the GPU path: our Pallas kernel has the ES kernel hardcoded inside, so it can't take a custom one. I'm open to a proposal, but we need the dispatch to route to Pallas only when no custom kernel is passed, and send custom kernels to the pure-JAX path.

**Comment by michael-0brien** [comment]
Ah makes sense. The pallas calls are a big reason of why I’d benefit from the API, and looking at the code it seems challenging to come up with a way to support this generically. Avoiding materializing the spreading arrays is really fantastic!

Let me think about how difficult this would be and weigh against if we should implement something custom.

**Comment by michael-0brien** [comment]
Hi @geoffroyO, I took a look at things more closely and naively it seems not too difficult to put together a PR that would include the pallas calls. 

Could you explain more why the `nufft.core.es_kernel` function isn’t called directly in the pallas spreading logic? There may be something I’m missing here. I’m not familiar with pallas yet, but I’m assuming all behavior under JIT will be the same, so passing around a kernel function wouldn’t be a problem.

**Comment by geoffroyO** [comment]
Hi @michael-0brien I've made a proposition in #11 !

**Comment by michael-0brien** [comment]
@geoffroyO this is awesome thank you so much for putting this together!

I was testing things out over the weekend and wasn’t able to get the pallas backend to work. I was using RTX Blackwell hardware, and was able to get unit tests passing, but for some reason during reverse-mode AD this error was being triggered within JAX: https://github.com/jax-ml/jax/blob/main/jax/_src/pallas/pallas_call.py#L271. Do you have a sense why this is happening?

In general I worry a bit about the stability of pallas within JAX. I wonder if, for now at least, it would make sense for users to opt in to the pallas backend. Would you consider creating a `NUFFTAX_PALLAS_SPREAD` environmental variable? This way nufftax by default relies on pure JAX and will be more robust to JAX version updates, different backends, etc. Here’s our model for setting an envar: https://github.com/michael-0brien/cryojax/blob/main/cryojax/_config.py

**Comment by geoffroyO** [comment]
Hi @michael-0brien  thanks again!

  Unfortunately I've only been able to test the Pallas backend on Hopper and Ampere — I don't have access to Blackwell
  GPUs, so I can't say for sure what's going wrong. My guess is that it's Pallas/Triton still being fragile on
  Blackwell.

  I completely agree on the env var. I was actually thinking of something a bit more general: a single
  `NUFFTAX_PALLAS_BACKEND` flag gating all the Pallas kernels (not just spread), defaulting to off so pure JAX stays the
  robust default. I'll put that together

**Comment by michael-0brien** [comment]
@geoffroyO this makes sense re Blackwell. I can test again for H100 at some point. To make sure it’s not something else and cover our bases: I was specifically taking reverse-mode gradients w.r.t non-uniform positions of a nufft2d1 call. Do we currently have GPU tests for this case under pallas?

And re the envar: that would be great! It seems like things are fast changing and the envar will also help alleviate some developer burden too. Nonetheless, these kernels are a great way to go and it looks like they’re here to stay in JAX.

Out of curiosity, have you tried messing around with pure JAX and inspecting the compiled HLO to see if it’s possible to nudge JIT into not materializing any O(nspread^d M) arrays? This can be tricky but I wonder if there are optimizations to be made there.

**Comment by geoffroyO** [comment]
Hi @michael-0brien, yes, the tests you're looking for are probably `test_nufft2d1_grad_xy_through_pallas_runs` and `test_spread_2d_grad_xy_through_pallas_finite_diff` they both pass on an A100 with the pallas backend active, so the issue is most likely specific to the Blackwell architecture.

On your last point: I haven't looked into it, but it's quite possible XLA is already doing that optimization ?

**Comment by michael-0brien** [comment]
Makes sense. I will investigate this further!

**Comment by michael-0brien** [comment]
Do you think it makes sense to set the pallas backend to “off” by default? This would make it so nufftax by default will always not throw errors, given the fast changing nature of pallas and differences across hardware. It would be awesome to turn it on by default maybe when pallas is moved out of jax.experimental? Even then I would lean towards keeping pure-JAX default, unless pallas does not restrict JAX transformation expressivity (e.g. multi-GPU).

It becomes complicated to take on nufftax as a dependency otherwise, as end-users of libraries like cryojax must have knowledge of nufftax configuration. What do you think?

Otherwise, there would need to be the ability to turn off pallas on-the-fly for downstream libraries, whereas envvars are read at import time.

**Comment by geoffroyO** [comment]
Makes sense. I defaulted it to true because I use it in many other projects, but you’re right, I’ll change it 😅

**Comment by michael-0brien** [comment]
Lol no problem, I relate to this. I appreciate you doing this!
